EMULATOR/LIBGPU.C

```diff
@@ -470,92 +529,148 @@ void DrawSplit(const VertexBufferSplit &split)
 	}
 }
 
-void DrawOTagEnv(u_long* p, DRAWENV* env)
+//
+// Draws all polygons after AggregatePTAG
+//
+void DrawAggregatedSplits()
 {
-	Emulator_BeginScene();
-#if defined(DEBUG_POLY_COUNT)
-	polygon_count = 0;
-#endif
+	// next code ideally should be called before EndScene
+	Emulator_UpdateVertexBuffer(g_vertexBuffer, g_vertexIndex);
 
-	PutDrawEnv(env);
+	for (int i = 1; i <= g_splitIndex; i++)
+		DrawSplit(g_splits[i]);
 
-	if (activeDrawEnv.isbg)
+	ClearVBO();
+}
+
+// forward declarations
+int ParsePrimitive(uintptr_t primPtr);
+int ParseLinkedPrimitiveList(uintptr_t packetStart, uintptr_t packetEnd);
+
+void AggregatePTAGsToSplits(u_long* p, bool singlePrimitive)
+{
+	if (!p)
+		return;
+
+	if (singlePrimitive)
 	{
-		ClearImage(&activeDrawEnv.clip, activeDrawEnv.r0, activeDrawEnv.g0, activeDrawEnv.b0);
-	} else {
-		Emulator_BlitVRAM();
+		// single primitive
+		ParsePrimitive((uintptr_t)p);
+		g_splits[g_splitIndex].vCount = g_vertexIndex - g_splits[g_splitIndex].vIndex;
 	}
-
-	if (p != NULL)
+	else
 	{
-		g_vertexIndex = 0;
-		g_splitIndex = 0;
-		g_splits[g_splitIndex].primType = PT_NONE; // first split is dummy
-
 		P_TAG* pTag = (P_TAG*)p;
 
-		do
+		// P_TAG as primitive list
+		//do
+		while ((uintptr_t)pTag != (uintptr_t)&terminator)
 		{
 			if (pTag->len > 0)
 			{
-				ParseLinkedPrimitiveList((uintptr_t)pTag, (uintptr_t)pTag + (uintptr_t)(pTag->len * 4) + 4 + LEN_OFFSET);
+				int lastSize = ParseLinkedPrimitiveList((uintptr_t)pTag, (uintptr_t)pTag + (uintptr_t)(pTag->len * 4) + 4 + LEN_OFFSET);
+				if (lastSize == -1)
```

Comment:
I'm unsure of this, the last primitive address in the OT is always -1 so it should always successfully terminate the loop.

Reply:
It's part of a safety feature to not get into infinite loop if you didn't properly set up primitive. Mostly my problems was caused by USE_32_BIT_ADDR=0, but it's good to have this feature

Reply:
It might not if you still have unsupported primitive types